lib/reqd.ml

```diff
@@ -138,29 +136,32 @@ let respond_with_bigstring t response (bstr:Bigstringaf.t) =
   if t.error_code <> `Ok then
     failwith "httpaf.Reqd.respond_with_bigstring: invalid state, currently handling error";
   match t.response_state with
-  | Waiting when_done_waiting ->
+  | Waiting ->
     (* XXX(seliopou): check response body length *)
     Writer.write_response     t.writer response;
     Writer.schedule_bigstring t.writer bstr;
     if t.persistent then
       t.persistent <- Response.persistent_connection response;
     t.response_state <- Complete response;
-    done_waiting when_done_waiting
+    Writer.wakeup t.writer;
   | Streaming _ ->
     failwith "httpaf.Reqd.respond_with_bigstring: response already started"
   | Complete _ ->
     failwith "httpaf.Reqd.respond_with_bigstring: response already complete"
 
 let unsafe_respond_with_streaming ~flush_headers_immediately t response =
   match t.response_state with
-  | Waiting when_done_waiting ->
+  | Waiting ->
     let response_body = Body.create t.response_body_buffer in
     Writer.write_response t.writer response;
-    if not flush_headers_immediately then Writer.yield t.writer;
     if t.persistent then
       t.persistent <- Response.persistent_connection response;
     t.response_state <- Streaming(response, response_body);
-    done_waiting when_done_waiting;
+    Body.when_ready_to_write response_body (fun () ->
+      Writer.wakeup t.writer);
+    if flush_headers_immediately
+    then Writer.wakeup t.writer
+    else Writer.yield t.writer;
```

Comment:
It looks like the async test you added is sufficient to justify not calling `wakeup` even when the headers shouldn't be flushed. Just wanted to confirm that's your understanding too.

Reply:
I think I misread your comment at first. We essentially could have made this change before by writing:

```ocaml
if flush_headers_immediately
then Writer.wakeup t.writer
else done_waiting when_done_waiting;
```

instead of always calling `done_waiting`. I think this is the correct behavior, and I think the old way would just force the writer to wake up only to immediately yield. But because I misread your comment, I went an added another test for the `wakeup` case that wasn't being exercised (i.e., if  you replaced the `then` branch with `()`, the tests would have passed).